### ISSUE TITLE
Support pruned branches in the CFG

### ISSUE THREAD

**Issue by Xazax-hun** [mgehre/clang#18]
Right now we turned on trivially false branch pruning.
This might be good for developer convenience and performance but we need to deal with in in some cases:
```
int *p = true? q : nullptr;
```

In the case above we should not even query and merge the pset of the false branch for ternary operator. Otherwise we and up with p pointing to invalid, since nullptr was never visited in the linearized cfg. This is worse than having null in the pset.

**Comment by mgehre** [comment]
Can we check that an `Expr` is trivially true/false (just what the CFG Builder does when pruning those branches)? If we have that, that we could use this is VisitConditionalExpr to only look at the non-pruned sub expression.

**Comment by Xazax-hun** [comment]
We could replicate the logic from the CFG builder but after a quick check I think that logic is rather convoluted and expensive. They even cache the results to avoid doing repeated evaluation of the same AST snippet. AnalysisDeclContext can build a CfgStmtMap. We could use that to look up the block for the corresponding Stmt and check if the block was visited. What do you think?

**Comment by mgehre** [comment]
I had hoped that the CFG builder just uses https://clang.llvm.org/doxygen/ExprConstant_8cpp_source.html

**Comment by Xazax-hun** [comment]
Ultimately it does, but do other things as well and I have no idea why: https://clang.llvm.org/doxygen/CFG_8cpp_source.html#l01100

**Comment by mgehre** [comment]
This would be a good question for the cfe mailing list

**Comment by Xazax-hun** [comment]
Ok, I think I know why the CFG has more convoluted logic for this.
Imagine the following example:
```
if ( a || b || true )
```

If we just evaluate the condition, we will return true. But in the CFG we have several branches due to the short circuit.
Se the CFG needs to make sure it will not just prune all the edges regarding this condition but make sure to evaluate `a` and `b` and only prune the branch regarding the last operand.

I think we do not really want to replicate this in the lifetime checker.

Checking if the block of a statement was visited is pretty easy in the LifetimeContext object. That object, however, is not accessible in the LifetimePsetBuilder.cpp

Since we already pass lots of the LifetimeContext's members to the LifetimePSetBuilder, I think it might make sense to pass a reference to the whole context instead.  For this, we might need to move LifetimeContext out into a header file. Are you ok with this direction?

**Comment by mgehre** [comment]
Would it be enough to see if there is no entry in ExprPsets and then skip
that subexpr?

Gábor Horváth <notifications@github.com> schrieb am Fr., 7. Sep. 2018,
17:56:

**Comment by Xazax-hun** [comment]
I am not sure. Sometimes we set pset to empty deliberately to avoid further warnings. Maybe if we set Psets to static in those cases? Also we would need to get rid of interpreting "no pset" for an expression as temporary. So I think this might not be an easy change.

**Comment by mgehre** [comment]
When we set the pset to empty (=unknown), we at least have an entry in the
subexprpset map. Don't we assert on "no pset" currently?

Gábor Horváth <notifications@github.com> schrieb am Fr., 7. Sep. 2018,
18:32:

**Comment by Xazax-hun** [comment]
This is what `getPSet` looks like currently:
```
  PSet getPSet(const Expr *E) {
    E = IgnoreTransparentExprs(E);
    if (E->isLValue()) {
      auto I = RefersTo.find(E);
      assert(I != RefersTo.end());
      return I->second;
    } else {
      auto I = PSetsOfExpr.find(E);
      if (I == PSetsOfExpr.end())
        return PSet::singleton(Variable::temporary());
      return I->second;
    }
  }
```

Basically if we did not set the Pset for an rvalue, we just say that it must be a temporary that we never saw. I do not remember what is the reason we had this special case but turning this into an assertion like `assert(I != PSetsOfExpr.end() || !hasPSet(E))` makes all of our test cases fail. 

I think the basic intention might have been something like this: we do not want to create a PSet for all the 250+ AST nodes explicitly, we rather fall back to producing a temporary.

**Comment by mgehre** [comment]
Fixed in 6d2cdc0a
